File: 46 Permutations/Python/solution.py
from typing import List, Self


class Solution:
    def backtrack(self: Self, nums: List[int], path: List[int], permutations: List[List[int]]) -> None:
        if len(path) == len(nums):
            permutations.append(path.copy())
        
        for num in nums:
            if num not in path:
                path.append(num)
                self.backtrack(nums, path, permutations)
                path.pop()

    def permute(self: Self, nums: List[int]) -> List[List[int]]:
        permutations: List[List[int]] = []
        self.backtrack(nums, [], permutations)
        return permutations

File: 46 Permutations/Java/Solution.java
import java.util.ArrayList;
import java.util.List;

public class Solution {
    private static void backtrack(int[] nums, List<List<Integer>> permutations, List<Integer> permutation) {
        if (permutation.size() == nums.length) {
            permutations.add(new ArrayList<>(permutation));
            return;
        }

        for (int num : nums) {
            if (!permutation.contains(num)) {
                permutation.add(0, num);
                backtrack(nums, permutations, permutation);
                permutation.remove(0);
            }
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        List<List<Integer>> permutations = new ArrayList<>();
        backtrack(nums, permutations, new ArrayList<>());
        return permutations;
    }
}
